crypto/bn/asm/rsaz-4k-avx512.pl

```diff
@@ -0,0 +1,939 @@
+#
+# The input |red_table| contains precomputations for two independent base values,
+# so the |tbl_idx| indicates for which base shall we extract the value.
+# |red_table_idx| is a power index.
+#
+# Extracted value (output) is 40 digit number in 2^52 radix.
+#
+# void extract_multiplier_2x40_win5(BN_ULONG *red_Y,
+#                                   const BN_ULONG red_table[1 << EXP_WIN_SIZE][2][40],
+#                                   int red_table_idx,
+#                                   int tbl_idx);           # 0 or 1
+#
+# EXP_WIN_SIZE = 5
+###############################################################################
+{
+# input parameters
+my ($out,$red_tbl,$red_tbl_idx,$tbl_idx)=$win64 ? ("%rcx","%rdx","%r8", "%r9") :  # Win64 order
+                                                  ("%rdi","%rsi","%rdx","%rcx");  # Unix order
+
+my ($t0,$t1,$t2,$t3,$t4,$t5) = map("%ymm$_", (0..5));
+my ($t6,$t7,$t8,$t9) = map("%ymm$_", (16..19));
+my $t0xmm = $t0 =~ s/%y/%x/r;
+my ($tmp0,$tmp1,$tmp2,$tmp3,$tmp4,$tmp5,$tmp6,$tmp7,$tmp8,$tmp9) = map("%ymm$_", (20..29));
+my ($cur_idx,$idx) = map("%ymm$_", (30..31));
+
+$code.=<<___;
+.text
+
+.align 32
+.globl  extract_multiplier_2x40_win5
+.type   extract_multiplier_2x40_win5,\@abi-omnipotent
+extract_multiplier_2x40_win5:
+.cfi_startproc
+    endbranch
+    leaq    ($tbl_idx,$tbl_idx,4), %rax
+    salq    \$6, %rax
+    addq    %rax, $red_tbl
+
+    vpbroadcastq    $red_tbl_idx, $idx
+    leaq   `(1<<5)*2*40*8`($red_tbl), %rax  # holds end of the tbl
+
+    vpxor   $t0xmm, $t0xmm, $t0xmm
+    vmovdqa64   $t0, $t9
+    vmovdqa64   $t0, $t8
+    vmovdqa64   $t0, $t7
+    vmovdqa64   $t0, $t6
+    vmovdqa64   $t0, $t5
+    vmovdqa64   $t0, $t4
+    vmovdqa64   $t0, $t3
+    vmovdqa64   $t0, $t2
+    vmovdqa64   $t0, $t1
+    vmovdqa64   $t0, $cur_idx
+
+.align 32
+.Lloop:
+    vpcmpq  \$0, $cur_idx, $idx, %k1            # mask of (idx == cur_idx)
+    addq    \$`2*40*8`, $red_tbl
+    vpaddq  .Lones(%rip), $cur_idx, $cur_idx    # increment cur_idx
+    vmovdqu64  `-2*40*8+0*32`($red_tbl), $tmp0     # load data from red_tbl
+    vmovdqu64  `-2*40*8+1*32`($red_tbl), $tmp1
+    vmovdqu64  `-2*40*8+2*32`($red_tbl), $tmp2
+    vmovdqu64  `-2*40*8+3*32`($red_tbl), $tmp3
+    vmovdqu64  `-2*40*8+4*32`($red_tbl), $tmp4
+    vmovdqu64  `-2*40*8+5*32`($red_tbl), $tmp5
+    vmovdqu64  `-2*40*8+6*32`($red_tbl), $tmp6
+    vmovdqu64  `-2*40*8+7*32`($red_tbl), $tmp7
+    vmovdqu64  `-2*40*8+8*32`($red_tbl), $tmp8
+    vmovdqu64  `-2*40*8+9*32`($red_tbl), $tmp9
+    vpblendmq  $tmp0, $t0, ${t0}{%k1}           # extract data when mask is not zero
+    vpblendmq  $tmp1, $t1, ${t1}{%k1}
+    vpblendmq  $tmp2, $t2, ${t2}{%k1}
+    vpblendmq  $tmp3, $t3, ${t3}{%k1}
+    vpblendmq  $tmp4, $t4, ${t4}{%k1}
+    vpblendmq  $tmp5, $t5, ${t5}{%k1}
+    vpblendmq  $tmp6, $t6, ${t6}{%k1}
+    vpblendmq  $tmp7, $t7, ${t7}{%k1}
+    vpblendmq  $tmp8, $t8, ${t8}{%k1}
+    vpblendmq  $tmp9, $t9, ${t9}{%k1}
```

Comment:
same comment as for extract_multiplier_2x20_win5 re using writemask with vmovdqu64 applies here too

Reply:
addressed with other similar comment resolution